### ISSUE TITLE
Local package dependency resolution failure

### ISSUE THREAD

**Issue by krzyzanowskim** [swiftlang/swift-package-manager#7889]
### Is it reproducible with SwiftPM command-line tools: `swift build`, `swift test`, `swift package` etc?

- [X] Confirmed reproduction steps with SwiftPM CLI. The description text _must_ include reproduction steps with either of command-line SwiftPM commands, `swift build`, `swift test`, `swift package` etc.

### Bug Report: Dependency Resolution Failure

**Issue**:
When attempting to add the following package: [STTextView-Plugin-Neon](https://github.com/krzyzanowskim/STTextView-Plugin-Neon) as a dependency, the dependency resolution fails with the error message: 
```
Failed to resolve dependencies. Dependencies could not be resolved because root depends on 'sttextview-plugin-neon' 0.5.0..<1.0.0.
```

**Screenshots**:
![Screenshot 2024-08-15 at 12 54 59](https://github.com/user-attachments/assets/2cdd7177-674b-460c-b275-35f35e7aa382)
![Error Details](https://github.com/user-attachments/assets/96cdf995-87f1-4c26-92e0-ad7e9f48108e)

**Possible Cause**:
It appears that there's a circular dependency issue. The package might be inadvertently creating a cycle by depending on itself, preventing successful resolution.

**Hypothesis**:
The problem could be related to a local package dependency `.package(path: "TreeSitterLanguages")`. Although this is a different package identity, the resolution process should theoretically handle this without errors (in my opinion).

**Steps to Reproduce**:
1. Add the STTextView-Plugin-Neon package as a dependency in your project.
2. Attempt to resolve dependencies.

**Expected Result**:
Dependencies should resolve without errors.

**Actual Result**:
Dependency resolution fails with the error specified above.

Looking forward to a resolution for this issue. Thank you!

### Expected behavior

Package dependency resolution does not break

### Actual behavior

Failed to resolve dependencies Dependencies could not be resolved because root depends on 'sttextview-plugin-neon' 0.5.0..<1.0.0.

### Steps to reproduce

1. Create empty Xcode project for macOS app
2. Add `https://github.com/krzyzanowskim/STTextView-Plugin-Neon` package dependency

### Swift & OS version (output of `swift --version ; uname -a`)

Swift 5.10
23.6.0 Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:04 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T8122 arm64

Confirmed with Xcode 15, Xcode 16 beta5

**Comment by plemarquand** [comment]
This error occurred because STTextView-Plugin-Neon releases published <=0.5.0 had a dependency on TreeSitterLanguages defined as a local package, i.e: `.package(path: "TreeSitterLanguages")`. When pulling it SPM doesn't know where to find TreeSitterLanguages locally and it fails. 

If you want to develop against a local version of the package you can publish releases that use a public release and then use `swift package edit` to specify an overlay that points to your local version.

**Comment by krzyzanowskim** [comment]
> When pulling it SPM doesn't know where to find TreeSitterLanguages locally and it fails.

I disagree. It does know. It is in the same repository and SPM pull the whole repository. The path exists and is accessible. It doesn't use that context.

> and then use swift package edit to specify an overlay

yes, but. that doesn't work well with Xcode workflow

**Comment by bripeticca** [comment]
SwiftPM seems to handle local dependencies differently when loading a remote dependency's manifest (as opposed to the root package's local dependency, which it _does_ handle appropriately), and fails to consider searching within its repository filesystem.

In my search, I have found that it will assess whether the dependencies for a remote **versioned** dependency are **also versioned**, and if not it will fail - as @plemarquand mentioned, this dependency on STTextView-Plugin-Neon is defined as `.package(path: "TreeSitterLanguages")` which (for good reason! 😄) is not versioned, and currently SwiftPM will throw an error when it encounters this scenario.

 I have noticed that SwiftPM does retain the context for the remote filesystem, so it (at one point) has the necessary context to be able to find the local dependency. I'm currently taking a look at this!
